chia/wallet/wallet_node.py

```diff
@@ -957,175 +960,180 @@ async def new_peak_wallet(self, new_peak: wallet_protocol.NewPeakWallet, peer: W
+        )
 
-                    # For every block, we need to apply the cache from race_cache
-                    for potential_height in range(backtrack_fork_height + 1, new_peak.height + 1):
-                        header_hash = self.wallet_state_manager.blockchain.height_to_hash(uint32(potential_height))
-                        if header_hash in self.race_cache:
-                            self.log.info(f"Receiving race state: {self.race_cache[header_hash]}")
-                            await self.receive_state_from_peer(list(self.race_cache[header_hash]), peer)
+        if new_peak_hb.height < self.constants.WEIGHT_PROOF_RECENT_BLOCKS:
+            # this is the case happens chain is shorter then WEIGHT_PROOF_RECENT_BLOCKS
+            await self.sync_from_untrusted_close_to_peak(new_peak_hb, peer)
+            return
 
-                    self.wallet_state_manager.state_changed("new_block")
-                    self.wallet_state_manager.set_sync_mode(False)
-                    self.log.info(f"Finished processing new peak of {new_peak.height}")
+        if not far_behind and peer.peer_node_id in self.synced_peers:
+            # This is the (untrusted) case where we already synced and are not too far behind. Here we just
+            # fetch one by one.
+            await self.sync_from_untrusted_close_to_peak(new_peak_hb, peer)
+            return
 
-        if peer.peer_node_id in self.synced_peers:
-            await self.wallet_state_manager.blockchain.set_finished_sync_up_to(new_peak.height)
+        # we haven't synced fully to this peer yet
+        syncing = False
+        if far_behind or len(self.synced_peers) == 0:
+            syncing = True
+            self.wallet_state_manager.set_sync_mode(True)
+
+        secondery_sync_running = (
+            self._secondary_peer_sync_task is not None and self._secondary_peer_sync_task.done() is False
+        )
+        if not syncing and secondery_sync_running:
+            self.log.info("Will not do secondary sync, there is already another sync task running.")
+            return
+
+        if await self.check_for_synced_trusted_peer(new_peak_hb, request_time):
+            self.log.info("Cancelling untrusted sync, we are connected to a trusted peer")
+            return
+
+        try:
+            await self.long_sync_from_untrusted(syncing, new_peak_hb, peer)
+        except Exception as e:
+            tb = traceback.format_exc()
+            self.log.error(f"Error syncing to {peer.get_peer_info()} {e} {tb}")
+            tb = traceback.format_exc()
+            self.log.error(f"Error syncing to {peer.get_peer_info()} {tb}")
+            await peer.close()
+            if syncing:
+                self.wallet_state_manager.set_sync_mode(False)
+        return
+
+    async def long_sync_from_untrusted(self, syncing: bool, new_peak_hb: HeaderBlock, peer: WSChiaConnection):
```

Comment:
I take it the code was moved verbatim into these functions, right? i.e. this is a pure refactor, no change in logic.

Reply:
i removed some unreachable code and avoid a WP request where it not needed
avoid a block header request for the peak in the case where we don't use it and added validation that the peak matches the block we requested (also added a test for that) which resulted in unexpected behaviour by the node


the intention was for the logic to stay the same, having said that this code was not easy to refactor.